### ISSUE TITLE
PayerInfo.getBillingAddress() is null for approved PayPal payments

### ISSUE THREAD

**Issue by dormitionskete** [paypal/PayPal-Java-SDK#102]
We are doing a new PayPal Express integration for an ecommerce store that needs to retrieve the Billing and Shipping Address information after the user logs into PayPal to authorize the payment, but BEFORE we capture the payment.

We need that information so we can create a user account in our store for them, if one does not already exist for them, and so we can properly calculate the shipping charges for the customer.

The following post to the StackOverflow forum over a year ago implied that had not been implemented into the REST API yet. Is that still the case?

http://stackoverflow.com/questions/29910936/is-paypal-rest-api-still-lacking-an-equivalent-of-getexpresscheckoutdetails

If it has been implemented, would somebody please point out to me how to access that information? If so, I would greatly appreciate it.

If it has not been implemented, does anyone know when it will be implemented?

**Comment by jziaja** [comment]
After a `payment` resource has been approved by a buyer, you can get the the addresses for the buyer by doing the following:

````java
Payment payment = Payment.get(apiContext, paymentId);

ShippingAddress shippingAddress = payment.getPayer().getPayerInfo().getShippingAddress();
Address billingAddress = payment.getPayer().getPayerInfo().getBillingAddress();
````

I just tested this, and the API is setting this information correctly.

**Comment by dormitionskete** [comment]
Thank you very much.  This *almost* works for me.

I’m getting the shipping address with this, but it is returning a null for the billing address.

With the Classic API, I had to set the "REQBILLINGADDRESS = 1" in the config files, and send that to PayPal with the initial request that sent the user to the PayPal site.  I had PayPal configure our Sandbox account to allow that to be returned.  It works fine with the Classic API.

With the REST API, I don’t see any way of sending the information to request the billing address information to be returned with the payment information.  

Is there something special I need to do to get it to return the billing address?

Again, any help will be *greatly* appreciated!

**Comment by jziaja** [comment]
Thanks for the explanation, @dormitionskete.  I've forwarded this question to the Payments API team to see if there's a reason why the `billing_address` information isn't being returned.  I'll follow up with you here once I hear back from them.

(Reopening and changing issue title to reflect the missing `payer_info` information.)

**Comment by dormitionskete** [comment]
Thank you very much!  I really appreciate it!

**Comment by fuvizzo** [comment]
The issue persists. I faced the same problem today when I implemented REST API in my application and I needed to retrieve billing info once payment has been executed.

I wonder PayPal team has nothing to tell us about that issue.
Come on guys let us know something.

Thanks!

**Comment by dormitionskete** [comment]
On Jun 1, 2015, at 5:17 PM, fuvizzo <notifications@github.com> wrote:

> The issue persists. I faced the same problem today when I implemented REST API in my application and I needed to retrieve billing info once payment has been executed.
> 
> I wonder PayPal team has nothing to tell us about that issue.
> Come on guys let us know something.
> 
> Thanks!
> 
> —
> Reply to this email directly or view it on GitHub.
> 


Well, I put in a support ticket with the PayPal Merchant Technical Services about this, and this is what I got back:

Their reply:

Hi,
Thank you for contacting Merchant Technical Services.
As much as I would like to help you, currently it is intended that the billing_address information isn't being returnedin REST API. My sincere apologies for the inconveniences caused due to the limitations in REST API. However, the developers are working hard to make this feature available in future. Kindly be informed that the Classic API is not going to be deprecated in the near future.
Thanks.
Sincerely,
Sheela
Merchant Technical Support
PayPal, an eBay Company

To access the whole contact:

To access your question from our support site, click here.

The bottom line is, it looks like anyone needing that functionality needs to use the Classic API.  

Hope this helps!

**Comment by fuvizzo** [comment]
I CANNOT BELIEVE. 

Pure science fiction.
Neither George Lucas nor Steven Spielberg could do better.

**Comment by ppmtscory** [comment]
This feature should be available as long as you have the option to receive the Billing Address turned on in your PayPal account.  With the REST API, you won't need to pass REQBILLINGADDRESS, you simply need the option to receive the billing information turned on in your account.  In order to get this option turned on, you must contact our customer support folks.  You can reach them via email or phone by clicking the 'contact us' link at the bottom of any PayPal page.

**Comment by dormitionskete** [comment]
I had contacted PayPal support to enable the Billing Address for our account before I opened this ticket, and it never would work; so if you haven't made any changes to implement this feature in the REST API, then it's still not working.

**Comment by ppmtscory** [comment]
@dormitionskete it works just fine for me when it's turned on in my account.  It's possible that the feature was added after you opened this ticket last May.  Using the SDK samples, I added the following for when I'm returned back after payment approval...
```
Payment payment1 = Payment.get(accessToken, req.getParameter("paymentId"));
LOGGER.info("Payment retrieved ID = " + req.getParameter("paymentId") + ", Billing Address = " + payment1.getPayer().getPayerInfo().getBillingAddress());
```
The result from the logger is...
```
 INFO [1311253681@qtp-1013199775-0] (PaymentWithPayPalServlet.java:204) - Created payment with id = PAY-4BA294891B725721DK2F63SA and status = created
 INFO [1311253681@qtp-1013199775-0] (PaymentWithPayPalServlet.java:120) - Payment retrieved ID = PAY-4BA294891B725721DK2F63SA, Billing Address = {
  "line1": "123 main st",
  "line2": "",
  "city": "BEVERLY HILLS",
  "country_code": "US",
  "postal_code": "90210",
  "state": "CA"
}
```

**Comment by dormitionskete** [comment]
Thank you very much for your response.

This seems to work for my Live credentials now, but it does not work with my test credentials in the Sandbox environment.  

I have submitted a ticket to get this turned on in my Sandbox environment.

I appreciate your help very much!

**Comment by dormitionskete** [comment]
Below is the response I just got back from PayPal support.

Do you know anyone at PayPal Support that you can talk with to help get this straightened out?


Subject
Retrieve PayPal Express Billing Address in SANDBOX Environment.
 
 Discussion Thread
 Response Via Email (Vicaci)	01/06/2016 12:28 AM
Dear, ___________,

Thank you for taking the time to contact us with your concerns. My name is Vicaci and I am the technical support specialist in Paypal. 

For REST API, billing address is only returned when the payment_method is the credit card type.
So for Paypal as the payment_method, the shipping address will be returned instead of billing address.
This is the product limitation for REST API. Our product team will expend all the features in future.
Please kindly keep turned for our product release.

In addition, Your sandbox account _________________________________ is already enabled for billing information which is only used for Classic API.

Thank you for your business and having a nice day!
 
 
Sincerely,  
Vicaci Zhu
 
Merchant Technical Support
PayPal

**Comment by ppmtscory** [comment]
@dormitionskete I see that you are conversing with Vicaci in the ticket and it's been updated since this last post.  I double checked the sandbox account that you stated in the ticket and it does, in fact, have the billing address option enabled on it already.  Therefore, please ensure that you are using the correct client credentials.  If needed, please update the ticket in response to Vicaci's last request.  If you need additional assistance, please let me know.

**Comment by dormitionskete** [comment]
Dear Cory,

Do you have a Java sample project that you could send me that works?  I’m getting nowhere fast with this.  I’m just spinning my wheels with it.  

I think that I need to start over from scratch with something that works.  

Vicaci says my test credentials work for him, so it’s a problem with my code, not the sandbox environment.  

If you have something you could send me, I’d *greatly* appreciate it.

Cordially,

Peter, hieromonk

**Comment by ppmtscory** [comment]
@dormitionskete ya, it's definitely not your credentials as I just tested with those and am getting the billing address back just fine.  I'm just using the Samples included in the SDK.  After you change the [sdk_config.properties](https://github.com/paypal/PayPal-Java-SDK/blob/master/rest-api-sample/src/main/resources/sdk_config.properties) file to use your credentials, then the instructions for running the samples can be found [here](https://github.com/paypal/PayPal-Java-SDK/tree/master/rest-api-sample).  You'll see that the billing address comes back in the Execute response when running the complete sample of a PayPal payment.  If you want to get it prior to executing, then you'll need to put the GET call for the Payment ID on your return URL so that it's grabbed prior to you running the Execute call.

If you are still unable to see it by using the SDK samples, please let me know and we can try to do a screen share to see what may be going wrong.

**Comment by dormitionskete** [comment]
Ok. This is what I have done:

Resynced my PayPal-Java-SDK project with GitHub.

Replaced rest-api-sample and rest-api-sdk folders in a Netbeans test project that does not have any other work of my own in it.  It is like a fresh project, as if it was straight from GitHub, except I might have added some dependencies last year to get it to compile.  But otherwise, it's fresh from GitHub, with no code of my own in it.

Replaced the clientID and clientSecret in the rest-api-sample/src/main/resources/sdk_config.properties file with our own credentials for our DsWebStore02 application.

Cleaned and built the project.

Ran the project.

Changed the url from localhost to https://trinity.gocamerica.net:8443/rest-api-sample/ and refreshed the page so PayPal can return to it.

Pressed the "Try It" button on the top item on the list called "PayPal Payments - similar to Express Checkout in Classic APIs".

That took me to https://trinity.gocamerica.net:8443/rest-api-sample/paymentwithpaypal

That gave me the following Request Object:

Request Object
```
{
  "intent": "sale",
  "payer": {
    "payment_method": "paypal"
  },
  "transactions": [
    {
      "transactions": [],
      "amount": {
        "currency": "USD",
        "total": "7",
        "details": {
          "shipping": "1",
          "subtotal": "5",
          "tax": "1"
        }
      },
      "description": "This is the payment transaction description.",
      "item_list": {
        "items": [
          {
            "name": "Ground Coffee 40 oz",
            "quantity": "1",
            "price": "5",
            "currency": "USD"
          }
        ]
      }
    }
  ],
  "redirect_urls": {
    "return_url": "https://trinity.gocamerica.net:8443/rest-api-sample/paymentwithpaypal?guid=15c065d1cc05411c910f4b7fa3df614d",
    "cancel_url": "https://trinity.gocamerica.net:8443/rest-api-sample/paymentwithpaypal?guid=15c065d1cc05411c910f4b7fa3df614d"
  }
}
```
And the Response was:

Response Object
```
{
  "id": "PAY-66992101DE771910UK2HS5FQ",
  "intent": "sale",
  "state": "created",
  "payer": {
    "payment_method": "paypal"
  },
  "transactions": [
    {
      "amount": {
        "total": "7.00",
        "currency": "USD",
        "details": {
          "subtotal": "5.00",
          "tax": "1.00",
          "shipping": "1.00"
        }
      },
      "description": "This is the payment transaction description.",
      "item_list": {
        "items": [
          {
            "name": "Ground Coffee 40 oz",
            "price": "5.00",
            "currency": "USD",
            "quantity": 1
          }
        ]
      },
      "related_resources": []
    }
  ],
  "create_time": "2016-01-08T03:35:49Z",
  "links": [
    {
      "href": "https://api.sandbox.paypal.com/v1/payments/payment/PAY-66992101DE771910UK2HS5FQ",
      "rel": "self",
      "method": "GET"
    },
    {
      "href": "https://www.sandbox.paypal.com/cgi-bin/webscr?cmd=_express-checkout&token=EC-08K865542M086990B",
      "rel": "approval_url",
      "method": "REDIRECT"
    },
    {
      "href": "https://api.sandbox.paypal.com/v1/payments/payment/PAY-66992101DE771910UK2HS5FQ/execute",
      "rel": "execute",
      "method": "POST"
    }
  ]
}
```
I pressed the link labeled "Redirect to PayPal to approve the payment".

That took me to PayPal.

I logged into PayPal using "DsWebStore-buyer@GOCAmerica.net", and pressed the "Continue" button.  

That took me back to https://trinity.gocamerica.net:8443/rest-api-sample/paymentwithpaypal?guid=1e9ecc99c5dd4b66aca4ddb26a9e1561&paymentId=PAY-488659069M758735JK2HSD2I&token=EC-2MB7316602913283E&PayerID=GBKLADMBQQW9E

This was my Request object:

Request Object
```
{
  "payer_id": "GBKLADMBQQW9E"
}
```
This was my Response:

Response Object
```
{
  "id": "PAY-66992101DE771910UK2HS5FQ",
  "intent": "sale",
  "state": "approved",
  "cart": "08K865542M086990B",
  "payer": {
    "payment_method": "paypal",
    "payer_info": {
      "email": "DsWebStore-buyer@GOCAmerica.net",
      "first_name": "test",
      "last_name": "buyer",
      "payer_id": "GBKLADMBQQW9E",
      "shipping_address": {
        "recipient_name": "Jane Doe",
        "line1": "123 Nowhere Street",
        "line2": "Apt. 1",
        "city": "La Crosse",
        "state": "KS",
        "postal_code": "67548",
        "country_code": "US"
      },
      "phone": "4086026235",
      "country_code": "US"
    }
  },
  "transactions": [
    {
      "amount": {
        "total": "7.00",
        "currency": "USD",
        "details": {
          "subtotal": "5.00",
          "tax": "1.00",
          "shipping": "1.00"
        }
      },
      "description": "This is the payment transaction description.",
      "item_list": {
        "items": [
          {
            "name": "Ground Coffee 40 oz",
            "price": "5.00",
            "currency": "USD",
            "quantity": 1
          }
        ],
        "shipping_address": {
          "recipient_name": "Jane Doe",
          "line1": "123 Nowhere Street",
          "line2": "Apt. 1",
          "city": "La Crosse",
          "state": "KS",
          "postal_code": "67548",
          "country_code": "US"
        }
      },
      "related_resources": [
        {
          "sale": {
            "id": "8XL395411F2665607",
            "state": "completed",
            "amount": {
              "total": "7.00",
              "currency": "USD",
              "details": {
                "subtotal": "5.00",
                "tax": "1.00",
                "shipping": "1.00"
              }
            },
            "payment_mode": "INSTANT_TRANSFER",
            "protection_eligibility": "ELIGIBLE",
            "protection_eligibility_type": "ITEM_NOT_RECEIVED_ELIGIBLE,UNAUTHORIZED_PAYMENT_ELIGIBLE",
            "transaction_fee": {
              "value": "0.50",
              "currency": "USD"
            },
            "parent_payment": "PAY-66992101DE771910UK2HS5FQ",
            "create_time": "2016-01-08T03:39:38Z",
            "update_time": "2016-01-08T03:39:38Z",
            "links": [
              {
                "href": "https://api.sandbox.paypal.com/v1/payments/sale/8XL395411F2665607",
                "rel": "self",
                "method": "GET"
              },
              {
                "href": "https://api.sandbox.paypal.com/v1/payments/sale/8XL395411F2665607/refund",
                "rel": "refund",
                "method": "POST"
              },
              {
                "href": "https://api.sandbox.paypal.com/v1/payments/payment/PAY-66992101DE771910UK2HS5FQ",
                "rel": "parent_payment",
                "method": "GET"
              }
            ]
          }
        }
      ]
    }
  ],
  "create_time": "2016-01-08T03:39:38Z",
  "links": [
    {
      "href": "https://api.sandbox.paypal.com/v1/payments/payment/PAY-66992101DE771910UK2HS5FQ",
      "rel": "self",
      "method": "GET"
    }
  ]
}
```
This was the output from my Netbeans IDE:
```
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
 INFO [http-nio-8443-exec-19] (PaymentWithPayPalServlet.java:199) - Created payment with id = PAY-66992101DE771910UK2HS5FQ and status = created
```
---

I did not change any code in any of the files except the clientID and clientSecret in the properties file.

As you can see, it did not return the Billing Address.

This is not my code.  This is your code.  Well, not necessarily your's personally, but it is code that came straight from GitHub, so it is code written by somebody on the PayPal-Java-SDK team, or PayPal staff, or at least somebody that knows a lot more about this than I do.  And it is good sample code for us to work with as a base.

---

Ideally, what we need is to be able to retrieve both the Shipping Address and the Billing Address immediately upon the customer authorizing the payment and returning to our site, but before applying the payment.  If we cannot retrieve both addresses before the payment is charged to their PayPal Account, we at least need to be able to retrieve the Shipping Address so we can calculate the shipping charges, and give the customer the ability to select how they would like their purchase shipped.

Then we need to update the payment amount to reflect the Purchase + Shipping prior to charging their PayPal Account.  For testing purposes, just add a couple of dollars to the price.

If we are not able to retrieve the Billing Address until after the payment is actually made, then we at least need to get it after the payment has been made so we can create an appropriate record for them in our WebStore's database.

---

So, would it be possible for you to take this sample code and modify it accordingly to do that?  I assume it would be the PaymentWithPayPalServlet.java class that would need to be modified.  (?)

If you could do that for us, please, I think we would be in business!  It'd be a great help, and I would really appreciate it.

Otherwise, why in the world do you get the Billing Address, but we do not?

And am I missing a major step here, that is needed to get the Billing Address?  

I'd *greatly* appreciate any help you can give us with this.

Thank you very much for your patience with me up to this point, too!

**Comment by ppmtscory** [comment]
@dormitionskete thanks for that.  I can now see what's going on and it's absolutely nothing that you are doing wrong.  After comparing your transaction and mine, I can see that you are getting redirected through an older checkout flow, whereas, when I do it (even with your creds) I get directed through the newer checkout flow.  *WHY* this is happening; I will have to find out.  

However, do me a favor real quick.  Run the same test, but when you get the redirect URL, don't just click it but copy the link and paste it in the browser.  Then append `#/checkout/login` to the end of it and hit enter.  This should send you through the newer checkout flow and you should see the billing address in the end.

This also explains why you see it in Live but not Sandbox using the same code.

**Comment by dormitionskete** [comment]
That did not seem to work for me.  However, just to be sure that I did what you wanted, this is what I did:

Instead of clicking on the "Redirect to PayPal to approve the payment" link, I copied it, pasted it into the browser and added what you wanted.  The resulting URL was:  

https://www.sandbox.paypal.com/cgi-bin/webscr?cmd=_express-checkout&token=EC-3LG4621257947803A#/checkout/login

That took me to:

https://www.sandbox.paypal.com/cgi-bin/webscr?cmd=_express-checkout&token=EC-3LG4621257947803A#pageState=review&pageDispatch=50a222a57771920b6a3d7b606239e4d529b525e0b7e69bf0224adecfb0124e9b61f737ba21b081984719ecfa9a8ffe80733a1a700ced90ae&pageSession=IJO8S8wQ5GK70-UApeGj20IOrY8H-6eDcfjdCGf-8I0W1xZZ3NdzvzgYDr4

It returned me to:

https://trinity.gocamerica.net:8443/rest-api-sample/paymentwithpaypal?guid=ab1dc37fcaf64a3295a887417f93065f&paymentId=PAY-9XR436168C1205006K2HUY5Y&token=EC-3LG4621257947803A&PayerID=GBKLADMBQQW9E

The response was:

```
{
  "id": "PAY-9XR436168C1205006K2HUY5Y",
  "intent": "sale",
  "state": "approved",
  "cart": "3LG4621257947803A",
  "payer": {
    "payment_method": "paypal",
    "payer_info": {
      "email": "DsWebStore-buyer@GOCAmerica.net",
      "first_name": "test",
      "last_name": "buyer",
      "payer_id": "GBKLADMBQQW9E",
      "shipping_address": {
        "recipient_name": "Mary Lou Wheeler",
        "line1": "712 E. Loomis St.",
        "city": "McPherson",
        "state": "KS",
        "postal_code": "67460",
        "country_code": "US"
      },
      "phone": "4086026235",
      "country_code": "US"
    }
  },
  "transactions": [
    {
      "amount": {
        "total": "7.00",
        "currency": "USD",
        "details": {
          "subtotal": "5.00",
          "tax": "1.00",
          "shipping": "1.00"
        }
      },
      "description": "This is the payment transaction description.",
      "item_list": {
        "items": [
          {
            "name": "Ground Coffee 40 oz",
            "price": "5.00",
            "currency": "USD",
            "quantity": 1
          }
        ],
        "shipping_address": {
          "recipient_name": "Mary Lou Wheeler",
          "line1": "712 E. Loomis St.",
          "city": "McPherson",
          "state": "KS",
          "postal_code": "67460",
          "country_code": "US"
        }
      },
      "related_resources": [
        {
          "sale": {
            "id": "3T4681235N3172644",
            "state": "completed",
            "amount": {
              "total": "7.00",
              "currency": "USD",
              "details": {
                "subtotal": "5.00",
                "tax": "1.00",
                "shipping": "1.00"
              }
            },
            "payment_mode": "INSTANT_TRANSFER",
            "protection_eligibility": "ELIGIBLE",
            "protection_eligibility_type": "ITEM_NOT_RECEIVED_ELIGIBLE,UNAUTHORIZED_PAYMENT_ELIGIBLE",
            "transaction_fee": {
              "value": "0.50",
              "currency": "USD"
            },
            "parent_payment": "PAY-9XR436168C1205006K2HUY5Y",
            "create_time": "2016-01-08T05:45:33Z",
            "update_time": "2016-01-08T05:45:33Z",
            "links": [
              {
                "href": "https://api.sandbox.paypal.com/v1/payments/sale/3T4681235N3172644",
                "rel": "self",
                "method": "GET"
              },
              {
                "href": "https://api.sandbox.paypal.com/v1/payments/sale/3T4681235N3172644/refund",
                "rel": "refund",
                "method": "POST"
              },
              {
                "href": "https://api.sandbox.paypal.com/v1/payments/payment/PAY-9XR436168C1205006K2HUY5Y",
                "rel": "parent_payment",
                "method": "GET"
              }
            ]
          }
        }
      ]
    }
  ],
  "create_time": "2016-01-08T05:45:34Z",
  "links": [
    {
      "href": "https://api.sandbox.paypal.com/v1/payments/payment/PAY-9XR436168C1205006K2HUY5Y",
      "rel": "self",
      "method": "GET"
    }
  ]
}
```

Still no Billing Address.

Did I do what you wanted correctly?

**Comment by ppmtscory** [comment]
Yup, you did it correctly.  It looks like that appendage didn't actually take you through the newer checkout flow.  By chance are you doing this outside of the US?  The newer checkout flow very well may not be showing up due to the region you are initiating the calls from.  However, I can tell you with 100% certainty, that is what is causing you to not get the billing address.

**Comment by dormitionskete** [comment]
No.  I'm in America.  Colorado.

So, can you fix it?

Also, I just noticed that shipping address in the previous post probably shouldn't be plastered all over the internet.  If I edit that previous post to change the address, will there still be a history of it, or can you change it so it isn't shown as it is, please?

**Comment by ppmtscory** [comment]
I will need to check with some teams internally to see why you are still triggering the classic checkout flow and get back to you.

You should be able to edit your posts to 'xxx' out or remove the information that you want so that it's not visible to others visiting this thread in the future.

**Comment by dormitionskete** [comment]
I just edited the name and address.

**Comment by dormitionskete** [comment]
Great.  I really appreciate your help with this.

**Comment by dormitionskete** [comment]
Just FYI - We are using the PayPal Classic API in our WebStore.  

Last March, when I was told by somebody at PayPal that this functionality was not available in the Rest API, I switched over and developed the store using the Classic API.  Maybe that has something to do with it?

**Comment by ppmtscory** [comment]
@dormitionskete if you are using this SDK with a client ID and secret as your credentials, then you are using the REST API - and not the Classic API.  However, if you're asking if using the Classic APIs previously would cause this, then I say no that shouldn't be the case.  I'm still waiting to hear back from the team internally but will update you as soon as I do.

**Comment by dormitionskete** [comment]
I just wanted you to be aware that we are using the Classic API in our WebStore.  I don't know if it is a factor in the problem we are having, but I thought I should make that clear to you in case it might be contributing to the problem.

Also, you should know that we seem to have multiple problems with our test environment.  I opened another support ticket last night dealing with the fact that we can never see our transaction history or transaction detail for our DsWebStore-facilitator@GOCAmerica.net sandbox account.

I thought that since that is clearly out of your realm of responsibilities that I should open a new support ticket for that.

I just don't want whoever you have working on our issue, or the person working on these other sandbox issues, to clobber each other in your efforts to get our testing environment working properly.

I hope I didn't cause more trouble for you by doing that.  (?)  But I also thought we should probably try to get everything straightened out all at once.  I thought that would be a route that would be more likely to succeed.

Again, thank you very much for your assistance with all of this.  I appreciate it very much.

**Comment by dormitionskete** [comment]
Cory, this is the response that I received back tonight about our sandbox issues:

Thank you for the information. I have escalated this sandbox issue to our internal team and I will get back to you as soon as the issue has been fixed. Please be informed that in REST API, we can't always determine which checkout page will be return, classic or new one. We are currently enrolling all the merchants to the new check out page but there is no ETA on this. Therefore, sometimes you will still be directed to classic one.

What does this mean for us about our getting the REST API redirection problem that we have been working on resolved?

**Comment by ppmtscory** [comment]
@dormitionskete that's exactly the reason that you're having the other issue.  The reason you don't get the billing address via REST API is because you get the Classic view.  I've reached out to see if there's any way to make you always receive the new one, while the normal ramp of adding merchants to the new checkout continues, but I've not heard back yet.  Just know that this is, in fact, the reason it's not working for you and so once you receive the new checkout, you'll get the billing address back.  In the meantime, since it's working on Live, you could run a couple of transactions there for testing your code and then just refund the payments.

**Comment by dormitionskete** [comment]
Well, I'd rather you not break our Classic sandbox in order to get the Rest API to work.  We have another PayPal account.  If I was to create a developer account, new web app in the sandbox, etc., would that be pretty much guaranteed to get set up to go to the Rest view?

**Comment by ppmtscory** [comment]
Nothing is "guaranteed".. :)

But yes, if you wanted to login to developer and create a new app with that second account, you're more than welcome to try it.  Just FYI, though, even if you get the new checkout flow, your Classic API integration would still work the same.  The API functionality (of your classic integration) won't change - just the look of the checkout flow.

**Comment by dormitionskete** [comment]
No, I mean, have your folks not break our Classic environment that we both asked them to work on, and I'll create a the new account for the Rest environment.  Would that work?  And is it too late for that?

**Comment by ppmtscory** [comment]
I haven't heard back from the team yet on the classic vs new environment so no it's not too late.  You can try a new account and see if you get the new environment.  Just let me know what you see after testing.

**Comment by dormitionskete** [comment]
Will do.

**Comment by dormitionskete** [comment]
Well, your wonderful Sandbox isn't letting me create a merchant account for our other PayPal account.  I keep getting:  We’re sorry, something went wrong during account creation. Please try again.

**Comment by dormitionskete** [comment]
I have successfully created a new developer account, sandbox test merchant account, and test web application.  I have submitted support ticket #160112-000372 requesting the Billing Address to be returned on our PayPal Express Rest transactions.

Also, they seem to have fixed our Sandbox issues for our main account, except for the redirection problem.

I'll let you know what happens when I hear back that the Billing Address has been enabled.

Thank you again for all of your help and patience!  I do appreciate it very much.

**Comment by dormitionskete** [comment]
I was notified by PayPal that they had enabled us to receive the Billing Address info for the new developer account that I created today for this extra PayPal Account we have.

It does not work, either.  It appears to do the same thing!

Or am I mistaken?

And appending that code you asked me to try to the end of the redirect link did not help, either.

---

https://trinity.gocamerica.net:8443/rest-api-sample/paymentwithpaypal

Request Object
```
{
  "intent": "sale",
  "payer": {
    "payment_method": "paypal"
  },
  "transactions": [
    {
      "transactions": [],
      "amount": {
        "currency": "USD",
        "total": "7",
        "details": {
          "shipping": "1",
          "subtotal": "5",
          "tax": "1"
        }
      },
      "description": "This is the payment transaction description.",
      "item_list": {
        "items": [
          {
            "name": "Ground Coffee 40 oz",
            "quantity": "1",
            "price": "5",
            "currency": "USD"
          }
        ]
      }
    }
  ],
  "redirect_urls": {
    "return_url": "https://trinity.gocamerica.net:8443/rest-api-sample/paymentwithpaypal?guid=9c4ed76ef35e486bbd1794e4d900c7f0",
    "cancel_url": "https://trinity.gocamerica.net:8443/rest-api-sample/paymentwithpaypal?guid=9c4ed76ef35e486bbd1794e4d900c7f0"
  }
}
```
Response Object
```
{
  "id": "PAY-8MB362256Y881703UK2K3HRA",
  "intent": "sale",
  "state": "created",
  "payer": {
    "payment_method": "paypal"
  },
  "transactions": [
    {
      "amount": {
        "total": "7.00",
        "currency": "USD",
        "details": {
          "subtotal": "5.00",
          "tax": "1.00",
          "shipping": "1.00"
        }
      },
      "description": "This is the payment transaction description.",
      "item_list": {
        "items": [
          {
            "name": "Ground Coffee 40 oz",
            "price": "5.00",
            "currency": "USD",
            "quantity": 1
          }
        ]
      },
      "related_resources": []
    }
  ],
  "create_time": "2016-01-13T02:17:40Z",
  "links": [
    {
      "href": "https://api.sandbox.paypal.com/v1/payments/payment/PAY-8MB362256Y881703UK2K3HRA",
      "rel": "self",
      "method": "GET"
    },
    {
      "href": "https://www.sandbox.paypal.com/cgi-bin/webscr?cmd=_express-checkout&token=EC-65C66039N5875001U",
      "rel": "approval_url",
      "method": "REDIRECT"
    },
    {
      "href": "https://api.sandbox.paypal.com/v1/payments/payment/PAY-8MB362256Y881703UK2K3HRA/execute",
      "rel": "execute",
      "method": "POST"
    }
  ]
}
```

Redirect Link:

https://www.sandbox.paypal.com/cgi-bin/webscr?cmd=_express-checkout&token=EC-65C66039N5875001U

Request Object
```
{
  "payer_id": "EPKAHK24Z9MGL"
}
```
Response Object
```
{
  "id": "PAY-8MB362256Y881703UK2K3HRA",
  "intent": "sale",
  "state": "approved",
  "cart": "65C66039N5875001U",
  "payer": {
    "payment_method": "paypal",
    "payer_info": {
      "email": "BillWheeler@whee.com",
      "first_name": "Bill",
      "last_name": "Wheeler",
      "payer_id": "EPKAHK24Z9MGL",
      "shipping_address": {
        "recipient_name": "Paulette Farber",
        "line1": "123 Nowhere Street",
        "line2": "Apt. 231",
        "city": "Omaha",
        "state": "NE",
        "postal_code": "68114",
        "country_code": "US"
      },
      "phone": "4089215681",
      "country_code": "US"
    }
  },
  "transactions": [
    {
      "amount": {
        "total": "7.00",
        "currency": "USD",
        "details": {
          "subtotal": "5.00",
          "tax": "1.00",
          "shipping": "1.00"
        }
      },
      "description": "This is the payment transaction description.",
      "item_list": {
        "items": [
          {
            "name": "Ground Coffee 40 oz",
            "price": "5.00",
            "currency": "USD",
            "quantity": 1
          }
        ],
        "shipping_address": {
          "recipient_name": "Paulette Farber",
          "line1": "123 Nowhere Street",
          "line2": "Apt. 231",
          "city": "Omaha",
          "state": "NE",
          "postal_code": "68114",
          "country_code": "US"
        }
      },
      "related_resources": [
        {
          "sale": {
            "id": "3U99809918898482J",
            "state": "pending",
            "amount": {
              "total": "7.00",
              "currency": "USD",
              "details": {
                "subtotal": "5.00",
                "tax": "1.00",
                "shipping": "1.00"
              }
            },
            "payment_mode": "INSTANT_TRANSFER",
            "reason_code": "PAYMENT_REVIEW",
            "protection_eligibility": "INELIGIBLE",
            "transaction_fee": {
              "value": "0.50",
              "currency": "USD"
            },
            "parent_payment": "PAY-8MB362256Y881703UK2K3HRA",
            "create_time": "2016-01-13T02:19:41Z",
            "update_time": "2016-01-13T02:19:41Z",
            "links": [
              {
                "href": "https://api.sandbox.paypal.com/v1/payments/sale/3U99809918898482J",
                "rel": "self",
                "method": "GET"
              },
              {
                "href": "https://api.sandbox.paypal.com/v1/payments/sale/3U99809918898482J/refund",
                "rel": "refund",
                "method": "POST"
              },
              {
                "href": "https://api.sandbox.paypal.com/v1/payments/payment/PAY-8MB362256Y881703UK2K3HRA",
                "rel": "parent_payment",
                "method": "GET"
              }
            ]
          }
        }
      ]
    }
  ],
  "create_time": "2016-01-13T02:19:42Z",
  "links": [
    {
      "href": "https://api.sandbox.paypal.com/v1/payments/payment/PAY-8MB362256Y881703UK2K3HRA",
      "rel": "self",
      "method": "GET"
    }
  ]
}
```

**Comment by ppmtscory** [comment]
Yes, you are seeing the same thing.  All transactions coming into that account, I can tell, went through the classic flow.  I looked at your account, and billing address is enabled so the reason you aren't seeing it is due to getting the classic flow while using REST API.  Would you like me to follow up with the team internally again to see how we can get you on the new flow?

**Comment by dormitionskete** [comment]
Yes.  PLEASE escalate this to whatever level, to whatever person, that you can that can fix this.

Please!!!

Although I don't particularly like the idea of having to deal with two developer accounts, I think I'd rather you try to get somebody to fix this new account, because I really don't want our classic flow broken on our old one.

I still need our Classic API to work.  I just discovered a behavior in our store that I don't like, that I need to work on, so it's very important that our classic API continues to work properly.  But if you can get somebody to work on this new account, they probably aren't very likely to accidentally break anything on our old account.  They are completely different accounts, after all!

So yes, please, I'd *greatly* appreciate it if you could get *somebody* to fix this.  

We've been wrestling with this for over a week.  I'm sure you're probably getting tired of hearing from me, too!

**Comment by ppmtscory** [comment]
@dormitionskete I just wanted to update you that I provided your new account to the internal team and they are taking a look.

**Comment by ppmtscory** [comment]
@dormitionskete unfortunately I've been told that there's nothing that can 'whitelist' your account to get the newer checkout flow.  It's done via ramp process on our side and there's nothing within that ramp that can filter out anything account specific.  Simply know that this is the reason you are not getting billing address back and that it's not specific to your integration/code.  Since you're getting that checkout in Production, you should continue to be fine there.

**Comment by dormitionskete** [comment]
Thank you, Cory, for all of your help with this.  I really appreciate it.

I hope somebody at PayPal manages to get this situation straightened out before we (merchants) all have to switch over to the REST API.  It's kind of hard to do adequate testing without a properly functioning test environment!
